pravega/pravega-operator
# Issue 100: Delete PVC when scaling down

## Change log description

- Delete persistent volumes in bookie and segment store when scaling down.

- According to K8s documents, Pods are deleted in ordinal sequence when scaling down. 

- We just need to keep persistent volume claims with smaller ordinal number compared to replicas of statefulset.
## Purpose of the change
Fix #100

## How to test the change
Checkout to source branch, follow the instructions on the README.md to deploy a Pravega Cluster.

After you have a running Pravega Cluster, edit the replicas number in bookkeeper or segment store
```
$kubectl edit pravegacluster pravega
```
A yaml file will pop up, you can directly edit it with your text editor. Save the change and exit, k8s will automatically update your change.

Check Pod number and persistent volume claims number to see if they are matched with your change. In this case, I change the segment store replicas to 4. 
```
$kubectl get pod
NAME                                          READY   STATUS    RESTARTS   AGE
clunky-sasquatch-nfs-server-provisioner-0     1/1     Running   0          4d18h
example-0                                     1/1     Running   0          8m54s
example-1                                     1/1     Running   0          7m56s
example-2                                     1/1     Running   0          6m57s
pravega-bookie-0                              1/1     Running   0          11m
pravega-bookie-1                              1/1     Running   0          11m
pravega-bookie-2                              1/1     Running   0          11m
pravega-pravega-controller-576cfb874c-gpthz   1/1     Running   0          11m
pravega-pravega-segmentstore-0                1/1     Running   0          11m
pravega-pravega-segmentstore-1                1/1     Running   0          11m
pravega-pravega-segmentstore-2                1/1     Running   0          11m
pravega-pravega-segmentstore-3                0/1     Pending   0          2s
pravega-search-operator-6d999b8cbc-hzcvf      1/1     Running   0          3d1h
zookeeper-operator-789696cc4b-2k5r8           1/1     Running   0          4d18h


$kubectl get pvc
NAME                                   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
cache-pravega-pravega-segmentstore-0   Bound    pvc-92ac400a-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       13m
cache-pravega-pravega-segmentstore-1   Bound    pvc-92b01d1b-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       13m
cache-pravega-pravega-segmentstore-2   Bound    pvc-92bb6a7a-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       13m
cache-pravega-pravega-segmentstore-3   Bound    pvc-33030280-fc18-11e8-828a-42010a8000ad   20Gi       RWO            standard       114s
data-example-0                         Bound    pvc-dc6bfe6e-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       11m
data-example-1                         Bound    pvc-18f79f1f-fc17-11e8-828a-42010a8000ad   20Gi       RWO            standard       9m48s
data-example-2                         Bound    pvc-3c1379b9-fc17-11e8-828a-42010a8000ad   20Gi       RWO            standard       8m49s
journal-pravega-bookie-0               Bound    pvc-92380150-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       13m
journal-pravega-bookie-1               Bound    pvc-923eafcb-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       13m
journal-pravega-bookie-2               Bound    pvc-924a8278-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       13m
ledger-pravega-bookie-0                Bound    pvc-9238d7b9-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       13m
ledger-pravega-bookie-1                Bound    pvc-9241afe2-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       13m
ledger-pravega-bookie-2                Bound    pvc-9247d563-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       13m
pravega-tier2                          Bound    pvc-d6a3fa57-fc16-11e8-828a-42010a8000ad   50Gi       RWX            nfs            11m

```

Then use the same way to edit the replicas to scale down the cluster size. 
```
$kubectl edit pravegacluster pravega
$kubectl get pod
NAME                                          READY   STATUS    RESTARTS   AGE
clunky-sasquatch-nfs-server-provisioner-0     1/1     Running   0          4d19h
example-0                                     1/1     Running   0          40m
example-1                                     1/1     Running   0          39m
example-2                                     1/1     Running   0          38m
pravega-bookie-0                              1/1     Running   0          43m
pravega-bookie-1                              1/1     Running   0          43m
pravega-bookie-2                              1/1     Running   0          43m
pravega-pravega-controller-576cfb874c-gpthz   1/1     Running   0          43m
pravega-pravega-segmentstore-0                1/1     Running   0          43m
pravega-pravega-segmentstore-1                1/1     Running   0          43m
pravega-pravega-segmentstore-2                1/1     Running   0          43m
pravega-search-operator-6d999b8cbc-hzcvf      1/1     Running   0          3d1h
zookeeper-operator-789696cc4b-2k5r8           1/1     Running   0          4d19h

$kubectl get pvc
NAME                                   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
cache-pravega-pravega-segmentstore-0   Bound    pvc-92ac400a-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       14m
cache-pravega-pravega-segmentstore-1   Bound    pvc-92b01d1b-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       14m
cache-pravega-pravega-segmentstore-2   Bound    pvc-92bb6a7a-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       14m
data-example-0                         Bound    pvc-dc6bfe6e-fc16-11e8-828a-42010a8000ad   20Gi       RWO            standard       12m
data-example-1                         Bound    pvc-18f79f1f-fc17-11e8-828a-42010a8000ad   20Gi       RWO            standard       10m
data-example-2                         Bound    pvc-3c1379b9-fc17-11e8-828a-42010a8000ad   20Gi       RWO            standard       9m32s
journal-pravega-bookie-0               Bound    pvc-92380150-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       14m
journal-pravega-bookie-1               Bound    pvc-923eafcb-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       14m
journal-pravega-bookie-2               Bound    pvc-924a8278-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       14m
ledger-pravega-bookie-0                Bound    pvc-9238d7b9-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       14m
ledger-pravega-bookie-1                Bound    pvc-9241afe2-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       14m
ledger-pravega-bookie-2                Bound    pvc-9247d563-fc16-11e8-828a-42010a8000ad   10Gi       RWO            standard       14m
pravega-tier2                          Bound    pvc-d6a3fa57-fc16-11e8-828a-42010a8000ad   50Gi       RWX            nfs            12m
```

You can see the number of pvc in segment store is reduced to 3. 

Signed-off-by: wenqi Wenqi.Mou@emc.com
